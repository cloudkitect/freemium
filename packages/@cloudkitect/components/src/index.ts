export * from './ck-certificatemanager/ck-dns-validated-certificate';
export * from './ck-cloudfront/ck-distribution';
export * from './ck-core/ck-vendor-tags';
export * from './ck-core/ck-tag';
export * from './ck-core/ck-utils';
export * from './ck-core/ck-region';
export * from './ck-dynamodb/ck-table-v2';
export * from './ck-ec2/ck-vpc';
export * from './ck-ecr/ck-repository';
export * from './ck-ecs/ck-cluster';
export * from './ck-ecs/ck-fargate-clsuter';
export * from './ck-ecs/ck-fargate-service';
export * from './ck-ecs/ck-fargate-task-definition';
export * from './ck-elasticloadbalancingv2/ck-public-application-loadbalancer';
export * from './ck-s3/ck-bucket';
export * from './ck-stack/ck-stack';
