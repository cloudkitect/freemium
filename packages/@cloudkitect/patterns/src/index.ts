export * from './ck-website/ck-serverless-website';
export * from './ck-container/ck-serverless-container-app';