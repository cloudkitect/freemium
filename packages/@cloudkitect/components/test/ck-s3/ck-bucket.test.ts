/*
 * Copyright (c) CloudKitect Inc. All Rights Reserved.
 */

import { App } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { BlockPublicAccess } from 'aws-cdk-lib/aws-s3';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { CkAccountType, CkStack } from '../../lib';
import { CkBucket } from '../../src';

let app: App;
let stack: CkStack;
const stackProps = {
  ckAccountType: CkAccountType.PRODUCTION,
  ckApplication: 'Test',
  ckCompany: 'CloudKitect',
};

describe('Well Architected S3 Bucket Configurations', () => {
  beforeEach(() => {
    app = new App();
    stack = new CkStack(app, 'TestStack', stackProps);
  });

  test('Bucket default settings', () => {
    new CkBucket(stack, 'MyBucket', {});
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });

  const defaultEncryption = {
    ServerSideEncryptionConfiguration: [
      {
        ServerSideEncryptionByDefault: {
          SSEAlgorithm: 'AES256',
        },
      },
    ],
  };

  test.each`
      actual       | expected
      ${undefined} | ${defaultEncryption}
      ${null}      | ${defaultEncryption}
   `(
    'CC-59: bucket encryption status should be $expected when $actual',
    ({ actual, expected }) => {
      new CkBucket(stack, 'MyBucket', {
        encryption: actual,
      });

      const template = Template.fromStack(stack);
      template.hasResourceProperties('AWS::S3::Bucket', {
        BucketEncryption: expected,
      });
    },
  );

  test('CC-68: bucket public read access should be disabled by default', () => {
    new CkBucket(stack, 'MyBucket', {
      enforceSSL: false,
    });

    const template = Template.fromStack(stack);
    template.resourceCountIs('AWS::S3::BucketPolicy', 0);
  });

  test.each`
      actual       | expected
      ${undefined} | ${false}
      ${null}      | ${false}
      ${true}      | ${false}
   `(
    'CC-68: bucket secure transport should be $expected when $actual',
    ({ actual, expected }) => {
      new CkBucket(stack, 'MyBucket', {
        enforceSSL: actual,
      });

      const template = Template.fromStack(stack);
      template.hasResourceProperties('AWS::S3::BucketPolicy', {
        Bucket: {
          Ref: 'MyBucketF68F3FF0',
        },
        PolicyDocument: {
          Statement: Match.arrayWith([
            {
              Action: 's3:*',
              Condition: {
                Bool: {
                  'aws:SecureTransport': `${expected}`,
                },
              },
              Effect: 'Deny',
              Principal: {
                AWS: '*',
              },
              Resource: Match.anyValue(),
            },
          ]),
        },
      });
    },
  );

  test('CC-62: bucket public read access should have read bucket policy when enabled', () => {
    new CkBucket(stack, 'MyBucket', {
      publicReadAccess: true,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ACLS,
    });

    const template = Template.fromStack(stack);
    template.hasResourceProperties('AWS::S3::BucketPolicy', {
      PolicyDocument: {
        Statement: Match.arrayWith([
          {
            Action: 's3:GetObject',
            Effect: 'Allow',
            Principal: {
              AWS: '*',
            },
            Resource: Match.anyValue(),
          },
        ]),
      },
    });
  });

  const defaultAccess = {
    BlockPublicAcls: true,
    BlockPublicPolicy: true,
    IgnorePublicAcls: true,
    RestrictPublicBuckets: true,
  };
  const customAccess = {
    BlockPublicAcls: true,
    IgnorePublicAcls: true,
  };
  test.each`
      actual                             | expected
      ${undefined}                       | ${defaultAccess}
      ${null}                            | ${defaultAccess}
      ${BlockPublicAccess.BLOCK_ACLS} | ${customAccess}
   `(
    'CC-49: bucket public access should be $expected when $actual',
    ({ actual, expected }) => {
      new CkBucket(stack, 'MyBucket', {
        blockPublicAccess: actual,
      });
      const template = Template.fromStack(stack);
      template.hasResourceProperties('AWS::S3::Bucket', {
        PublicAccessBlockConfiguration: expected,
      });
    },
  );

  test.each`
      actual       | accountType
      ${undefined} | ${CkAccountType.PRODUCTION}
      ${null}      | ${CkAccountType.PRODUCTION}
      ${true}      | ${CkAccountType.PRODUCTION}
      ${true}      | ${CkAccountType.UAT}
   `(
    'CC-63: bucket versioning should be enabled when $actual in $accountType',
    ({ actual, accountType }) => {
      const myStack = new CkStack(app, 'MyStack', {
        ...stackProps,
        ckAccountType: accountType,
      });
      new CkBucket(myStack, 'MyBucket', {
        versioned: actual,
      });

      const template = Template.fromStack(myStack);
      template.hasResourceProperties('AWS::S3::Bucket', {
        VersioningConfiguration: {
          Status: 'Enabled',
        },
      });
    },
  );

  test.each`
      actual       | accountType
      ${undefined} | ${CkAccountType.UAT}
      ${null}      | ${CkAccountType.UAT}
      ${false}     | ${CkAccountType.UAT}
      ${false}     | ${CkAccountType.PRODUCTION}
   `(
    'CC-63: bucket versioning should be disabled when $actual in $accountType',
    ({ actual, accountType }) => {
      const myStack = new CkStack(app, 'MyStack', {
        ...stackProps,
        ckAccountType: accountType,
      });
      new CkBucket(myStack, 'MyBucket', {
        versioned: actual,
      });

      const template = Template.fromStack(myStack);
      template.hasResourceProperties('AWS::S3::Bucket', {
        VersioningConfiguration: Match.absent(),
      });
    },
  );
});
