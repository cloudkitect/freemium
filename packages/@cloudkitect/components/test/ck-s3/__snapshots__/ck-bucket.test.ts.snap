// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Well Architected S3 Bucket Configurations Bucket default settings 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyBucketF68F3FF0": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "CkComponents",
            "Value": "2.128.0",
          },
          {
            "Key": "CkVendor",
            "Value": "CloudKitect",
          },
        ],
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "MyBucketPolicyE7FBAC7B": {
      "Properties": {
        "Bucket": {
          "Ref": "MyBucketF68F3FF0",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "MyBucketF68F3FF0",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "MyBucketF68F3FF0",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
